tests/ManageCourses.Tests/ImporterTests/UcasDataMigratorTests.cs

```diff
@@ -106,120 +102,541 @@ public void ErroneousCourseLeavesOtherCoursesAlone()
+                .WithOptedIn()
+                .WithCourses(new List<Course>
+                {
+                    new CourseBuilder()
+                        .WithCode(courseCode)
+                        .WithName(unmodifiedCourseName)
+                        .WithAccreditingProvider(new ProviderBuilder()
+                            .WithOptedIn()
+                            .WithCode(accreditingProviderCode))
+                }));
+            Context.Providers.Add(new ProviderBuilder()
+                .WithOptedIn()
+                .WithCode(accreditingProviderCode2));
+            Context.Save();
+
+            UcasCourse ucasCourse = new PayloadCourseBuilder()
+                .WithInstCode(instCode)
+                .WithCrseCode(courseCode)
+                .WithAccreditingProvider(accreditingProviderCode2)
+                .WithCampusCode(campusCode);
+            var ucasPayload = new PayloadBuilder()
+                .WithInstitutions(new List<UcasInstitution>
+                {
+                    new PayloadInstitutionBuilder()
+                        .WithInstCode(instCode)
+                })
+                .WithCourses(new List<UcasCourse> {ucasCourse})
+                .WithCampuses(new List<UcasCampus>
+                {
+                    new PayloadCampusBuilder()
+                        .WithCampusCode(campusCode)
+                        .WithInstCode(instCode)
+                });
+
+            // act
+            DoImport(ucasPayload);
+
+            // assert
+            var provider = Context.Providers.Single(p => p.ProviderCode == instCode);
+            provider.Courses.Count.Should().Be(1);
+            provider.Courses.Single(c => c.CourseCode == courseCode)
+                .AccreditingProvider.ProviderCode.Should().Be(accreditingProviderCode);
+        }
+
+        [Test]
+        public void AllVariations()
+        {
+            // build a big payload with all of the above
+            const string instCode = "AA1";
+            const string instCodeOptedIn = "AA2";
+            const string instName = "Armadillo 1";
+            const string modifiedInstName = "Modified " + instName;
+            const string instNameOptedIn = "OptedIn Armadillo 2";
+            const string accreditingProviderCode = "AINST201";
+            const string accreditingProviderCodeOptedIn = "AINST202";
+            const string accreditingProviderCode3 = "AINST203";
+            const string accreditingProviderCode4 = "AINST204";
+            const string courseCode = "CRS101";
+            const string courseCode2 = "CRS102";
+            const string courseCode3 = "CRS103";
+            const string unmodifiedCourseName = "Primary";
+            const string modifiedCourseName = "Modified " + unmodifiedCourseName;
+            const string campusCode = "CAMP101";
+
+            // arrange
+            // accrediting providers
+            var accreditingProvider = new ProviderBuilder()
```

Comment:
Be explicit, that was why your last pr got rejected as there was little evidence that there was test coverage.

Noe that we have test coverage, the last thing we wanna do is revisit this to understand that this provider is opted out.